README.md
# hermez_sdk

[![pub package](https://img.shields.io/badge/pub-1.0.0-orange)](https://pub.dev/packages/hermez_sdk)
[![build](https://github.com/hermeznetwork/hermez_flutter_sdk/workflows/hermez_sdk/badge.svg)](https://github.com/hermeznetwork/hermez_flutter_sdk/actions?query=workflow%3Ahermez_sdk)
[![license](https://img.shields.io/badge/License-Apache-blue.svg)](https://github.com/hermeznetwork/hermez_mobile_library/blob/master/LICENSE)

## Description

This is a flutter Plugin for Hermez Mobile SDK (https://hermez.io). This plugin provides a cross-platform tool (iOS, Android) to communicate with the Hermez API and network.

## Installation

To use this plugin, add `hermez_sdk` as a [dependency](https://flutter.io/using-packages/) in your `pubspec.yaml` file like this

```yaml
dependencies:
  hermez_sdk: ^x.y.z
```
This will get you the latest version.

If you want to test a specific branch of the repository, pull `hermez_sdk` like this

```yaml
dependencies:
  hermez_sdk:
      git:
        url: ssh://git@github.com/hermeznetwork/hermez-flutter-sdk.git
        ref: branchPathName
```

Also, add the abi contracts in json files to the assets folder of your project and to your `pubspec.yaml` file like this

```yaml
assets:
    - HermezABI.json
    - ERC20ABI.json
    - WithdrawalDelayerABI.json
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/hermeznetwork/hermez_flutter_sdk/issues

## Setup

NOTE: In order to interact with Hermez, you will need to supply your own Ethereum node. You can check these links to help you set up a node (https://blog.infura.io/getting-started-with-infura-28e41844cc89, https://blog.infura.io/getting-started-with-infuras-ethereum-api).

## Usage

To start using this package first import it in your Dart file.

```dart
import 'package:hermez_sdk/hermez_sdk.dart';
```

### Initialization

To initialize the Hermez SDK you can call the init method with one of the supported environments as a parameter, or setup all the different parameters passing the environment 'custom'.

```dart
HermezSDK.init(
  'rinkeby',
   web3ApiKey: EXAMPLES_WEB3_API_KEY
);
```

or 

```dart
HermezSDK.init(
  'custom',
  envParams: EnvParams(
      EXAMPLES_WEB3_CHAIN_ID,
      {
        ContractName.hermez: EXAMPLES_HERMEZ_ROLLUP_ADDRESS, // Hermez
        ContractName.withdrawalDelayer:
            EXAMPLES_HERMEZ_WDELAYER_ADDRESS, // WithdrawalDelayer
      },
      EXAMPLES_HERMEZ_API_URL,
      EXAMPLES_HERMEZ_EXPLORER_URL,
      EXAMPLES_WEB3_URL + EXAMPLES_WEB3_API_KEY,
      EXAMPLES_WEB3_RDP_URL + EXAMPLES_WEB3_API_KEY),
);
```

### Supported Tokens

Before being able to operate on the Hermez Network, we must ensure that the token we want to operate with is listed. For that we make a call to the Hermez Coordinator API that will list all available tokens. All tokens in Hermez Network must be ERC20.

We can see there are 2 tokens registered. ETH will always be configured at index 0. The second token is HEZ. For the rest of the examples we will work with ETH. In the future, more tokens will be included in Hermez.

```dart
import 'package:hermez_sdk/api.dart' as coordinatorApi;
import 'package:hermez_sdk/model/tokens_response.dart';

...

Future<TokensResponse> getHermezSupportedTokens() async {
  TokensResponse tokensResponse = await coordinatorApi.getTokens();
  return tokensResponse;
}
```

```json
{
  "tokens": [
    {
      "itemId": 1,
      "id": 0,
      "ethereumBlockNum": 0,
      "ethereumAddress": "0x0000000000000000000000000000000000000000",
      "name": "Ether",
      "symbol": "ETH",
      "decimals": 18,
      "USD": 1787,
      "fiatUpdate": "2021-02-28T18:55:17.372008Z"
    },
    {
      "itemId": 2,
      "id": 1,
      "ethereumBlockNum": 8153596,
      "ethereumAddress": "0x2521bc90b4f5fb9a8d61278197e5ff5cdbc4fbf2",
      "name": "Hermez Network Token",
      "symbol": "HEZ",
      "decimals": 18,
      "USD": 5.365,
      "fiatUpdate": "2021-02-28T18:55:17.386805Z"
    }
  ],
  "pendingItems": 0
}
```

### Create Wallet

We can create a new Hermez wallet by providing the Ethereum private key of an Ethereum account. This wallet will store the Ethereum and Baby JubJub keys for the Hermez account. The Ethereum address is used to authorize L1 transactions, and the Baby JubJub key is used to authorize L2 transactions. We will create two wallets.

> [!NOTE]
> You will need to supply two private keys to test and initialize both accounts. The keys provided here are invalid and are shown as an example.

```dart
import 'package:hermez_sdk/hermez_wallet.dart';

...

void createHermezWallets() async {
    // load first account
    final wallet =
        await HermezWallet.createWalletFromPrivateKey(EXAMPLES_PRIVATE_KEY1);
    final HermezWallet hermezWallet = wallet[0];
    final String hermezEthereumAddress = wallet[1];

    // load second account
    final wallet2 =
        await HermezWallet.createWalletFromPrivateKey(EXAMPLES_PRIVATE_KEY2);
    final HermezWallet hermezWallet2 = wallet2[0];
    final String hermezEthereumAddress2 = wallet2[1];
}
```

### Move tokens from Ethereum to Hermez Network

Creating a Hermez account and depositing tokens is done simultaneously as an L1 transaction. In this example we are going to deposit 1 ETH tokens into the newly created Hermez accounts.

```dart
import 'package:hermez_sdk/tx.dart' as tx;
import 'package:hermez_sdk/utils.dart';
import 'package:hermez_sdk/hermez_compressed_amount.dart';

...

void moveTokensFromEthereumToHermez() async {
 
    // load  account and ethereum token

    ...

    // set amount to transfer
    final amount = 1.0;
    final amountDeposit = getTokenAmountBigInt(amount, tokenERC20.decimals);
    final compressedDepositAmount =
        HermezCompressedAmount.compressAmount(amountDeposit.toDouble());

    // perform deposit account 1
    String txHash = await tx.deposit(compressedDepositAmount, hermezEthereumAddress, tokenERC20,
        hermezWallet.publicKeyCompressedHex, EXAMPLES_PRIVATE_KEY1);
}
```

### Token Balance

A token balance can be obtained by querying the API and passing the hermezEthereumAddress of the Hermez account.

```dart
void getTokenBalance() async {

    // load  accounts and ethereum token
    
    ...

    // get sender account information
    final infoAccountSender = (await coordinatorApi
            .getAccounts(hermezEthereumAddress, [tokenERC20.id]))
        .accounts[0];

    // get receiver account information
    final infoAccountReceiver = (await coordinatorApi
            .getAccounts(hermezEthereumAddress2, [tokenERC20.id]))
        .accounts[0];
}
```

```json
[
  {
    "accountIndex": "hez:ETH:4253",
    "balance": "1099600000000000000",
    "bjj": "hez:dMfPJlK_UtFqVByhP3FpvykOg5kAU3jMLD7OTx_4gwzO",
    "hezEthereumAddress": "hez:0x74d5531A3400f9b9d63729bA9C0E5172Ab0FD0f6",
    "itemId": 4342,
    "nonce": 1,
    "token": {
      "USD": 1789,
      "decimals": 18,
      "ethereumAddress": "0x0000000000000000000000000000000000000000",
      "ethereumBlockNum": 0,
      "fiatUpdate": "2021-02-28T18:55:17.372008Z",
      "id": 0,
      "itemId": 1,
      "name": "Ether",
      "symbol": "ETH"
    }
  },
  {
    "accountIndex": "hez:ETH:4254",
    "balance": "1097100000000000000",
    "bjj": "hez:HESLP_6Kp_nn5ANmSGiOnhhYvF3wF5Davf7xGi6lwh3U",
    "hezEthereumAddress": "hez:0x12FfCe7D5d6d09564768d0FFC0774218458162d4",
    "itemId": 4343,
    "nonce": 6,
    "token": {
      "USD": 1789,
      "decimals": 18,
      "ethereumAddress": "0x0000000000000000000000000000000000000000",
      "ethereumBlockNum": 0,
      "fiatUpdate": "2021-02-28T18:55:17.372008Z",
      "id": 0,
      "itemId": 1,
      "name": "Ether",
      "symbol": "ETH"
    }
  }
]
```

We can see that the field accountIndex is formed by the token symbol it holds and an index. A Hermez account can only hold one type of token. Account indexes start at 256. Indexes 0-255 are reserved for internal use. Note that the balances do not match with the ammount deposited of 1 ETH because accounts already existed in Hermez Network before the deposit, so we performed a deposit on top instead.

Alternatively, an account query can be filtered using the assigned accountIndex

```dart
    final account1ByIdx = coordinatorApi.getAccount(infoAccountSender.accountIndex);

    final account2ByIdx = coordinatorApi.getAccount(infoAccountReceiver.accountIndex);

```

```json
[
  {
    "accountIndex": "hez:ETH:4253",
    "balance": "1099600000000000000",
    "bjj": "hez:dMfPJlK_UtFqVByhP3FpvykOg5kAU3jMLD7OTx_4gwzO",
    "hezEthereumAddress": "hez:0x74d5531A3400f9b9d63729bA9C0E5172Ab0FD0f6",
    "itemId": 4342,
    "nonce": 1,
    "token": {
      "USD": 1789,
      "decimals": 18,
      "ethereumAddress": "0x0000000000000000000000000000000000000000",
      "ethereumBlockNum": 0,
      "fiatUpdate": "2021-02-28T18:55:17.372008Z",
      "id": 0,
      "itemId": 1,
      "name": "Ether",
      "symbol": "ETH"
    }
  },
  {
    "accountIndex": "hez:ETH:4254",
    "balance": "1097100000000000000",
    "bjj": "hez:HESLP_6Kp_nn5ANmSGiOnhhYvF3wF5Davf7xGi6lwh3U",
    "hezEthereumAddress": "hez:0x12FfCe7D5d6d09564768d0FFC0774218458162d4",
    "itemId": 4343,
    "nonce": 6,
    "token": {
      "USD": 1789,
      "decimals": 18,
      "ethereumAddress": "0x0000000000000000000000000000000000000000",
      "ethereumBlockNum": 0,
      "fiatUpdate": "2021-02-28T18:55:17.372008Z",
      "id": 0,
      "itemId": 1,
      "name": "Ether",
      "symbol": "ETH"
    }
  }
]
```

### Move tokens from Hermez to Ethereum Network

Withdrawing funds is a two step process:

1. Exit
2. Withdrawal

#### Exit

The Exit transaction is used as a first step to retrieve the funds from Hermez Network back to Ethereum. There are two types of Exit transactions:

- Normal Exit, referred as Exit from now on. This is a L2 transaction type.
- Force Exit, an L1 transaction type which has extended guarantees that will be processed by the Coordinator. We will talk more about Force Exit here
    
The Exit is requested as follows:

```dart
void moveTokensFromHermezToEthereumStep1Exit() async {
    // load account

    ...
    
    // set amount to exit
    final amount = 0.0001;
    final amountExit = getTokenAmountBigInt(amount, tokenERC20.decimals);
    final compressedExitAmount =
    HermezCompressedAmount.compressAmount(amountExit.toDouble());

    // set fee in transaction
    final state = await coordinatorApi.getState();
    final userFee = state.recommendedFee.existingAccount;

    // generate L2 transaction
    final l2ExitTx = {
    'type': 'Exit',
    'from': infoAccountSender.accountIndex,
    'amount': compressedExitAmount,
    'fee': userFee
    };

    final exitResponse = await tx.generateAndSendL2Tx(l2ExitTx, hermezWallet, infoAccountSender.token);
  }
```

```json
{
  "status": 200,
  "id": "0x0257305cdc43060a754a5c2ea6b0e0f6e28735ea8e75d841ca4a7377aa099d91b7",
  "nonce": 2
}
```

After submitting our Exit request to the Coordinator, we can check the status of the transaction by calling the Coordinator's Transaction Pool. The Coordinator's transaction pool stores all those transactions that are waiting to be forged.

```dart
final txExitPool = await coordinatorApi.getPoolTransaction(exitResponse['id']);
```

```json
{
  "amount": "1000000000000000000",
  "fee": 204,
  "fromAccountIndex": "hez:ETH:4253",
  "fromBJJ": "hez:dMfPJlK_UtFqVByhP3FpvykOg5kAU3jMLD7OTx_4gwzO",
  "fromHezEthereumAddress": "hez:0x74d5531A3400f9b9d63729bA9C0E5172Ab0FD0f6",
  "id": "0x0257305cdc43060a754a5c2ea6b0e0f6e28735ea8e75d841ca4a7377aa099d91b7",
  "info": null,
  "nonce": 2,
  "requestAmount": null,
  "requestFee": null,
  "requestFromAccountIndex": null,
  "requestNonce": null,
  "requestToAccountIndex": null,
  "requestToBJJ": null,
  "requestToHezEthereumAddress": null,
  "requestTokenId": null,
  "signature": "38f23d06826be8ea5a0893ee67f4ede885a831523c0c626c102edb05e1cf890e418b5820e3e6d4b530386d0bc84b3c3933d655527993ad77a55bb735d5a67c03",
  "state": "pend",
  "timestamp": "2021-03-16T12:31:50.407428Z",
  "toAccountIndex": "hez:ETH:1",
  "toBjj": null,
  "toHezEthereumAddress": null,
  "token": {
    "USD": 1781.9,
    "decimals": 18,
    "ethereumAddress": "0x0000000000000000000000000000000000000000",
    "ethereumBlockNum": 0,
    "fiatUpdate": "2021-02-28T18:55:17.372008Z",
    "id": 0,
    "itemId": 1,
    "name": "Ether",
    "symbol": "ETH"
  },
  "type": "Exit"
}
```

We can see the state field is set to pend (meaning pending). There are 4 possible states:

1. pend : Pending
2. fging : Forging
3. fged : Forged
4. invl : Invalid
    
If we continue polling the Coordinator about the status of the transaction, the state will eventually be set to fged.

We can also query the Coordinator to check whether or not our transaction has been forged. getHistoryTransaction reports those transactions that have been forged by the Coordinator.

```dart
final txExitConf = await coordinatorApi.getHistoryTransaction(txExitPool.id);
```

And we can confirm our account status and check that the correct amount has been transfered out of the account.

```dart
final accountResponse = await coordinatorApi.getAccounts(hermezEthereumAddress, [tokenERC20.id]);
final infoAccount = accountResponse.accounts.length > 0 ? accountResponse.accounts[0]: null;
```

#### Withdraw

After doing any type of Exit transaction, which moves the user's funds from their token account to a specific Exit Merkle tree, one needs to do a Withdraw of those funds to an Ethereum L1 account. To do a Withdraw we need to indicate the accountIndex that includes the Ethereum address where the funds will be transferred, the amount and type of tokens, and some information to verify the ownership of those tokens. Additionally, there is one boolean flag. If set to true, the Withdraw will be instantaneous.

```dart
void moveTokensFromHermezToEthereumStep2Withdraw() async {
    // load ethereum token and account

    ...

    final exitInfoN = (await coordinatorApi.getExits(
            hermezEthereumAddress, true, tokenERC20.id))
        .exits;

    if (exitInfoN != null && exitInfoN.length > 0) {
      final exitInfo = exitInfoN.last;
      // set to perform instant withdraw
      final isInstant = true;

      // perform withdraw
      tx.withdraw(
          double.parse(exitInfo.balance),
          exitInfo.accountIndex,
          exitInfo.token,
          hermezWallet.publicKeyCompressedHex,
          exitInfo.batchNum,
          exitInfo.merkleProof.siblings,
          EXAMPLES_PRIVATE_KEY1,
          isInstant: isInstant);
    }
}
```

The funds should now appear in the Ethereum account that made the withdrawal.

#### Force Exit

This is the L1 equivalent of an Exit. With this option, the smart contract forces Coordinators to pick up L1 transactions before they pick up L2 transactions to ensure that L1 transactions will eventually be picked up.

This is a security measure. We don't expect users to need to make a Force Exit.

```dart
void moveTokensFromHermezToEthereumStep1ForceExit() async {
    // load ethereum token and account info
    
    ...

    // set amount to force exit
    final amount = 0.0001;
    final amountForceExit = getTokenAmountBigInt(amount, tokenERC20.decimals);
    final compressedForceExitAmount =
        HermezCompressedAmount.compressAmount(amountForceExit.toDouble());

    // perform force exit
    tx.forceExit(compressedForceExitAmount, infoAccountSender.accountIndex,
        tokenERC20, EXAMPLES_PRIVATE_KEY1);
}
```

The last step to recover the funds will be to send a new Withdraw request to the smart contract as we did after the regular Exit request.

```dart
void moveTokensFromHermezToEthereumStep2Withdraw() async {
    // load ethereum token and account

    ...

    final exitInfoN = (await coordinatorApi.getExits(
            hermezEthereumAddress, true, tokenERC20.id))
        .exits;

    if (exitInfoN != null && exitInfoN.length > 0) {
      final exitInfo = exitInfoN.last;
      // set to perform instant withdraw
      final isInstant = true;

      // perform withdraw

CHANGELOG.md
## 1.0.0+3

* changed license file
* minor fixes

## 1.0.0+2

* added metadata for pub.dev
* fixed readme file

## 1.0.0+1

* adds support for dart null safety
* formatted files

## 1.0.0

* adds support for iOS and Android 🎉 
* Hermez wallets creation
* complete communication with coordinators Api
* adds Deposit, Withdraw and Transfers operations


lib/tx_utils.dart
import 'dart:ffi';
import 'dart:io';
import 'dart:typed_data';

import 'package:flutter/services.dart';
import 'package:hermez_sdk/api.dart' as api;
import 'package:hermez_sdk/utils.dart';
import 'package:web3dart/crypto.dart';

import 'addresses.dart'
    show
        getAccountIndex,
        getEthereumAddress,
        isHermezAccountIndex,
        isHermezEthereumAddress,
        isHermezBjjAddress;
import 'environment.dart';
import 'fee_factors.dart' show feeFactors, feeFactorsAsBigInts;
import 'hermez_compressed_amount.dart';
import 'libs/circomlib.dart';
import 'model/token.dart';
import 'tx_pool.dart' show getPoolTransactions;

const String hermezPrefix = "hez:";

const Map<String, String> txType = {
  "Deposit": "Deposit",
  "CreateAccountDeposit": "CreateAccountDeposit",
  "Transfer": "Transfer",
  "Withdraw": "Withdrawn",
  "Exit": "Exit"
};

enum TxType { Deposit, CreateAccountDeposit, Transfer, Withdraw, Exit }

const Map<String, String> txState = {
  "Forged": "fged",
  "Forging": "fing",
  "Pending": "pend",
  "Invalid": "invl"
};

final DynamicLibrary nativeExampleLib = Platform.isAndroid
    ? DynamicLibrary.open("libbabyjubjub.so")
    : DynamicLibrary.process();

// 60 bits is the minimum bits to achieve enough precision among fee factor values < 192
// no shift value is applied for fee factor values >= 192
const bitsShiftPrecision = 60;

//final circomlib = CircomLib(lib: await SetupUtil.getDylibAsync());

/// Encodes the transaction object to be in a format supported by the Smart Contracts and Circuits.
/// Used, for example, to sign the transaction
///
/// @param {Object} transaction - Transaction object returned by generateL2Transaction
/// @param {String} providerUrl - Network url (i.e, http://localhost:8545). Optional
///
/// @returns {Object} encodedTransaction
Map<String, dynamic> encodeTransaction(Map<String, dynamic> transaction,
    {String? providerUrl}) {
  final Map<String, dynamic> encodedTransaction = Map.from(transaction);

  encodedTransaction["chainId"] = getCurrentEnvironment()!.chainId;

  encodedTransaction["fromAccountIndex"] =
      getAccountIndex(transaction["fromAccountIndex"]);
  if (transaction["toAccountIndex"] != null) {
    encodedTransaction["toAccountIndex"] =
        getAccountIndex(transaction["toAccountIndex"]);
  } else if (transaction["type"] == 'Exit') {
    encodedTransaction["toAccountIndex"] = 1;
  }

  if (transaction["toHezEthereumAddress"] != null) {
    encodedTransaction["toEthereumAddress"] =
        getEthereumAddress(transaction["toHezEthereumAddress"]);
  }

  return encodedTransaction;
}

/// Generates the L1 Transaction Id based on the spec
/// TxID (32 bytes) for L1Tx is the Keccak256 (ethereum) hash of:
/// bytes:   | 1 byte |             32 bytes                |
///                     SHA256(    8 bytes      |  2 bytes )
/// content: |  type  | SHA256([ToForgeL1TxsNum | Position ])
/// where type for L1UserTx is 0
/// @param {Number} toForgeL1TxsNum
/// @param {Number} currentPosition
///
/// @returns {String}
String getL1UserTxId(BigInt toForgeL1TxsNum, BigInt currentPosition) {
  final toForgeL1TxsNumBytes = Uint8List(8);
  final toForgeL1TxsNumView = ByteData.view(toForgeL1TxsNumBytes.buffer);
  toForgeL1TxsNumView.setUint64(0, toForgeL1TxsNum.toInt());

  final positionBytes = Uint8List(8);
  final positionView = ByteData.view(positionBytes.buffer);
  positionView.setUint64(0, currentPosition.toInt());

  final toForgeL1TxsNumHex =
      bytesToHex(toForgeL1TxsNumView.buffer.asUint8List());
  final positionHex =
      bytesToHex(positionView.buffer.asUint8List().sublist(6, 8));

  final v = toForgeL1TxsNumHex + positionHex;
  final h = bytesToHex(keccak256(hexToBytes(v)));

  return '0x00' + h;
}

/// Generates the Transaction Id based on the spec
/// TxID (33 bytes) for L2Tx is:
/// bytes:   | 1 byte |                    32 bytes                           |
///                     SHA256( 6 bytes | 4 bytes | 2 bytes| 5 bytes | 1 byte )
/// content: |  type  | SHA256([FromIdx | TokenID | Amount |  Nonce  | Fee    ])
/// where type for L2Tx is '2'
/// @param {Number} fromIdx - The account index that sends the transaction
/// @param {Number} tokenId - The tokenId being transacted
/// @param {Number} amount - The amount being transacted
/// @param {Number} nonce - Nonce of the transaction
/// @param {Number} fee - The fee of the transaction
/// @returns {String} Transaction Id
String getL2TxId(int fromIdx, int tokenId, double amount, int nonce, int fee) {
  final fromIdxBytes = Uint8List(8);
  final fromIdxView = ByteData.view(fromIdxBytes.buffer);
  fromIdxView.setUint64(0, fromIdx);

  final tokenIdBytes = Uint8List(8);
  final tokenIdView = ByteData.view(tokenIdBytes.buffer);
  tokenIdView.setUint64(0, tokenId);

  final amountF40 = HermezCompressedAmount.compressAmount(amount).value;
  final amountBytes = Uint8List(8);
  final amountView = ByteData.view(amountBytes.buffer);
  amountView.setUint64(0, BigInt.from(amountF40).toInt());

  final nonceBytes = Uint8List(8);
  final nonceView = ByteData.view(nonceBytes.buffer);
  nonceView.setUint64(0, nonce);

  final fromIdxHex = bytesToHex(fromIdxView.buffer.asUint8List().sublist(2, 8));
  final tokenIdHex = bytesToHex(tokenIdView.buffer.asUint8List().sublist(4, 8));
  final amountHex = bytesToHex(
      amountView.buffer.asUint8List().sublist(3, 8)); // float40: 5 bytes
  final nonceHex = bytesToHex(nonceView.buffer.asUint8List().sublist(3, 8));

  String feeHex = fee.toRadixString(16);
  if (feeHex.length == 1) {
    feeHex = '0' + feeHex;
  }

  final v = fromIdxHex + tokenIdHex + amountHex + nonceHex + feeHex;
  final h = bytesToHex(keccak256(hexToBytes(v)));

  return '0x02' + h;
}

/// Calculates the appropriate fee factor depending on what's the fee as a percentage of the amount
///
/// @param {Number} fee - The fee in token value
/// @param {BigInt} amount - The amount as a BigInt string
///
/// @return {Number} feeFactor
int getFeeIndex(num fee, num amount) {
  num low = 0;
  int mid;
  int high = feeFactors.length - 1;
  while (high - low > 1) {
    mid = ((low + high) / 2).floor();
    if (getFeeValue(mid, amount).toDouble() < fee) {
      low = mid;
    } else {
      high = mid;
    }
  }

  return high;
}

/// Compute fee in token value with an amount and a fee index
/// @param {Number} feeIndex - Fee selected among 0 - 255
/// @param {Number} amount - The amount of the transaction as a Scalar
/// @returns {BigInt} Resulting fee in token value
BigInt getFeeValue(num feeIndex, num amount) {
  if (feeIndex < 192) {
    final fee = BigInt.from(amount * feeFactorsAsBigInts[feeIndex as int]);
    return fee >> bitsShiftPrecision;
  } else {
    return BigInt.from(amount * feeFactorsAsBigInts[feeIndex as int]);
  }
}

/// Gets the transaction type depending on the information in the transaction object
/// If an account index is used, it will be 'Transfer'
/// If a Hermez address is used, it will be 'TransferToEthAddr'
/// If a BabyJubJub is used, it will be 'TransferToBjj'
///
/// @param {Object} transaction - Transaction object sent to generateL2Transaction
///
/// @return {String} transactionType
String getTransactionType(Map transaction) {
  if (transaction["to"] != null) {
    if (isHermezAccountIndex(transaction['to'])) {
      return 'Transfer';
    } else if (isHermezEthereumAddress(transaction['to'])) {
      return 'TransferToEthAddr';
    } else if (isHermezBjjAddress(transaction['to'])) {
      return 'TransferToBJJ';
    } else {
      return 'Transfer';
    }
  } else {
    return 'Exit';
  }
}

/// Calculates the appropriate nonce based on the current token account nonce and existing transactions in the Pool.
/// It needs to find the lowest nonce available as transactions in the pool may fail and the Coordinator only forges
/// transactions in the order set by nonces.
///
/// @param {Number} currentNonce - The current token account nonce returned by the Coordinator (optional)
/// @param {String} accountIndex - The account index
/// @param {String} bjj - The account's BabyJubJub
/// @param {Number} tokenId - The token id of the token in the transaction
///
/// @return {Number} nonce
Future<num?> getNonce(
    num? currentNonce, String? accountIndex, String bjj, num? tokenId) async {
  if (currentNonce != null) {
    return currentNonce;
  }

  final accountData = await api.getAccount(accountIndex!);
  var nonce = accountData.nonce;

  final List<dynamic> poolTxs = await getPoolTransactions(accountIndex, bjj);

  poolTxs.removeWhere((tx) => tx.token.id == tokenId);
  final List poolTxsNonces =
      poolTxs.where((tx) => tx.nonce).toList(); //map((tx) => tx.nonce);
  poolTxs.sort();

  // return current nonce if no transactions are pending
  if (poolTxsNonces.length > 0) {
    while (poolTxsNonces.indexOf(nonce) != -1) {
      nonce = nonce! + 1;
    }
  }

  return nonce;
}

/// Encode tx compressed data
/// @param {Object} tx - Transaction object
/// @returns {BigInt} Encoded tx compressed data
BigInt buildTxCompressedData(Map<String, dynamic> tx) {
  final signatureConstant = BigInt.parse('3322668559');
  BigInt res = BigInt.zero;
  final chainId =
      BigInt.from((tx['chainId'] != null ? tx['chainId'] : 0)) << 32;
  final fromAccountIndex = BigInt.from(
          (tx['fromAccountIndex'] != null ? tx['fromAccountIndex'] : 0)) <<
      48;
  final toAccountIndex =
      BigInt.from((tx['toAccountIndex'] != null ? tx['toAccountIndex'] : 0)) <<
          96;
  final tokenId =
      BigInt.from((tx['tokenId'] != null ? tx['tokenId'] : 0)) << 144;
  final nonce = BigInt.from((tx['nonce'] != null ? tx['nonce'] : 0)) << 176;
  final fee = BigInt.from((tx['fee'] != null ? tx['fee'] : 0)) << 216;
  final toBjjSign = BigInt.from((tx['toBjjSign'] != null ? 1 : 0)) << 224;

  res = res + signatureConstant; // SignConst --> 32 bits -> 4 bytes
  res = res + chainId; // chainId --> 16 bits
  res = res + fromAccountIndex; // fromIdx --> 48 bits
  res = res + toAccountIndex; // toIdx --> 48 bits
  res = res + tokenId; // tokenID --> 32 bits
  res = res + nonce; // nonce --> 40 bits
  res = res + fee; // userFee --> 8 bits
  res = res + toBjjSign; // toBjjSign --> 1 bit

  return res;
}

/// Build element_1 for L2HashSignature
/// @param {Object} tx - Transaction object returned by `encodeTransaction`
/// @returns {BigInt} element_1 L2 signature
BigInt buildElement1(Map<String, dynamic> tx) {
  BigInt res = BigInt.zero;

  final toEthereumAddress = BigInt.parse(
      tx['toEthereumAddress'] != null
          ? tx['toEthereumAddress'].substring(2)
          : '0',
      radix: 16);
  double amount = tx['amount'] != null ? double.parse(tx['amount']) : 0;
  final amountF = BigInt.from(
          HermezCompressedAmount.compressAmount(amount).value.toInt()) <<
      160;
  final maxNumBatch =
      BigInt.from(tx['maxNumBatch'] != null ? tx['maxNumBatch'] : 0) << 200;

  res = res + toEthereumAddress; // ethAddr --> 160 bits
  res = res + amountF; // amountF --> 40 bits
  res = res + maxNumBatch; // maxNumBatch --> 32 bits

  return res;
}

/// Builds the message to hash
///
/// @param {Object} encodedTransaction - Transaction object
///
/// @returns {BigInt} message to sign
BigInt buildTransactionHashMessage(Map<String, dynamic> encodedTransaction) {
  final BigInt txCompressedData = buildTxCompressedData(encodedTransaction);
  final element1 = buildElement1(encodedTransaction);
  final toBjjAy = encodedTransaction['toBjjAy'] != null
      ? (encodedTransaction['toBjjAy'].startsWith('0x')
          ? encodedTransaction['toBjjAy'].substring(2)
          : encodedTransaction['toBjjAy'])
      : '0';
  final rqTxCompressedDataV2 =
      encodedTransaction['rqTxCompressedDataV2'] != null
          ? (encodedTransaction['rqTxCompressedDataV2'].startsWith('0x')
              ? encodedTransaction['rqTxCompressedDataV2'].substring(2)
              : encodedTransaction['rqTxCompressedDataV2'])
          : '0';
  final rqToEthAddr = encodedTransaction['rqToEthAddr'] != null
      ? (encodedTransaction['rqToEthAddr'].startsWith('0x')
          ? encodedTransaction['rqToEthAddr'].substring(2)
          : encodedTransaction['rqToEthAddr'])
      : '0';

  final rqToBjjAy = encodedTransaction['rqToBjjAy'] != null
      ? (encodedTransaction['rqToBjjAy'].startsWith('0x')
          ? encodedTransaction['rqToBjjAy'].substring(2)
          : encodedTransaction['rqToBjjAy'])
      : '0';

  CircomLib circomLib = CircomLib();
  String hashPoseidon = circomLib.hashPoseidon(
      txCompressedData.toString(),
      element1.toString(),
      toBjjAy,
      rqTxCompressedDataV2,
      rqToEthAddr,
      rqToBjjAy);
  BigInt h = hexToInt(hashPoseidon);
  return h;
}

/// Prepares a transaction to be ready to be sent to a Coordinator.
///
/// @param {Object} transaction - ethAddress and babyPubKey together
/// @param {String} transaction.from - The account index that's sending the transaction e.g hez:DAI:4444
/// @param {String} transaction.to - The account index of the receiver e.g hez:DAI:2156. If it's an Exit, set to a falseable value
/// @param {String} transaction.amount - The amount being sent as a BigInt string
/// @param {Number} transaction.fee - The amount of tokens to be sent as a fee to the Coordinator
/// @param {Number} transaction.nonce - The current nonce of the sender's token account
/// @param {String} bJJ - The compressed BabyJubJub in hexadecimal format of the transaction sender
/// @param {Object} token - The token information object as returned from the Coordinator.
///
/// @return {Object} - Contains `transaction` and `encodedTransaction`. `transaction` is the object almost ready to be sent to the Coordinator. `encodedTransaction` is needed to sign the `transaction`

Future<Set<Map<String, dynamic>>> generateL2Transaction(
    Map tx, String bjj, Token token) async {
  final type = tx['type'] != null ? tx['type'] : getTransactionType(tx);
  final nonce = await getNonce(tx['nonce'], tx['from'], bjj, token.id);
  final toAccountIndex = isHermezAccountIndex(tx['to']) ? tx['to'] : null;
  final decompressedAmount =
      HermezCompressedAmount.decompressAmount(tx['amount']);
  final feeBigInt = getTokenAmountBigInt(tx['fee'], token.decimals!);

  String? toHezEthereumAddress;
  if (type == 'TransferToEthAddr') {
    toHezEthereumAddress = tx['to'];
  }
  /* else if (type == 'TransferToBJJ') {
    toHezEthereumAddress = tx['toAuxEthAddr'] != null
        ? tx['toAuxEthAddr']
        : INTERNAL_ACCOUNT_ETH_ADDR;
  }*/

  Map<String, dynamic> transaction = {};
  transaction.putIfAbsent('type', () => type);
  transaction.putIfAbsent('tokenId', () => token.id);
  transaction.putIfAbsent('fromAccountIndex', () => tx['from']);
  transaction.putIfAbsent('toAccountIndex',
      () => type == 'Exit' ? 'hez:${token.symbol}:1' : toAccountIndex);
  transaction.putIfAbsent('toHezEthereumAddress', () => toHezEthereumAddress);
  transaction.putIfAbsent(
      'toBjj', () => type == 'TransferToBJJ' ? tx['to'] : null);
  // Corrects precision errors using the same system used in the Coordinator
  transaction.putIfAbsent(
      'amount', () => decompressedAmount.toString().replaceAll(".0", ""));
  transaction.putIfAbsent(
      'fee', () => getFeeIndex(feeBigInt.toDouble(), decompressedAmount));
  transaction.putIfAbsent('nonce', () => nonce);
  transaction.putIfAbsent('requestFromAccountIndex', () => null);
  transaction.putIfAbsent('requestToAccountIndex', () => null);
  transaction.putIfAbsent('requestToHezEthereumAddress', () => null);
  transaction.putIfAbsent('requestToBjj', () => null);
  transaction.putIfAbsent('requestTokenId', () => null);
  transaction.putIfAbsent('requestAmount', () => null);
  transaction.putIfAbsent('requestFee', () => null);
  transaction.putIfAbsent('requestNonce', () => null);
  Map<String, dynamic> encodedTransaction = encodeTransaction(transaction);
  transaction.putIfAbsent(
      'id',
      () => getL2TxId(
          encodedTransaction['fromAccountIndex'],
          encodedTransaction['tokenId'],
          decompressedAmount,
          encodedTransaction['nonce'],
          encodedTransaction['fee']));
  return {transaction, encodedTransaction};
}
